### PR TITLE
Make swift::SourceLoc and swift::SourceRange usable as a key type for DenseSet/DenseMap.

### PR DESCRIPTION
**Pull request by mhong**
One use case is using DenseSet to manage a set of source locations, so that we
emit at most one compiler diagnostic per location (see
https://github.com/apple/swift/pull/18402).

### CODE REVIEW

**Comment by mhong** [comment]
@lattner Who would you suggest we loop in for review? Thanks.

File: include/swift/Basic/SourceLoc.h [hunk 1]
```diff
@@ -214,4 +215,31 @@ class CharSourceRange {
 
 } // end namespace swift
 
+namespace llvm {
+template <typename T> struct DenseMapInfo;
+
+template <> struct DenseMapInfo<swift::SourceRange> {
+  static swift::SourceRange getEmptyKey() { return swift::SourceRange(); }
```
Comment (lattner) [inline include/swift/Basic/SourceLoc.h]: I'd suggest changing this to:

SMLoc::getFromPointer(DenseMapInfo<const char *>::getEmptyKey())))

for consistency.

**Comment by lattner** [comment]
I can review it, LGTM.

**Review by lattner — APPROVED** [review]

**Comment by jrose-apple** [comment]
I do think you should use SourceLoc for your original test case, though. No reason to store twice as many pointers.

**Comment by mhong** [comment]
@jrose-apple, I tried out a simple example like:
```swift
public func test() {
  var a = 1.0
  a += 2.0
}
```

In the SIL code, I find that the insts that read the old `a` in `a += 2.0` and the insts that wrte to the new `a` in `a += 2.0` could share the same start loc, but have a different end loc.

e.g. I saw these two insts:
```
%4 = begin_access [modify] [static] %0 : $*Double

1: inst->getLoc().getSourceRange() = {Start = {Value = {
      Ptr = 0x877d061 "a += 2.0\n\n  let x = Tensor<Float>(1.0)\n}\n"}}, End = {Value = {
      Ptr = 0x877d061 "a += 2.0\n\n  let x = Tensor<Float>(1.0)\n}\n"}}}
```
vs
```
%5 = struct_element_addr %4 : $*Double, #Double._value

1: inst->getLoc().getSourceRange() = {Start = {Value = {
      Ptr = 0x877d061 "a += 2.0\n\n  let x = Tensor<Float>(1.0)\n}\n"}}, End = {Value = {
      Ptr = 0x877d066 "2.0\n\n  let x = Tensor<Float>(1.0)\n}\n"}}}
```
The former reads the old `a`, and the latter starts the computation for doing `+=`.

So it'd be safer to store the loc pair. We don't expect to generate a lot of sends/recvs warnings, but if performance ever becomes a concern, we can revisit this (e.g. maybe just warn on the first K occurrences globally, or within each function.)

**Comment by mhong** [comment]
@swift-ci please test

**Comment by jrose-apple** [comment]
I mean, even the range isn't unique per instruction, because of implicit expressions, or just AST-level expressions that take more than one instruction. You could also have unrelated diagnostics at the same location.

**Comment by mhong** [comment]
@jrose-apple Based on the feedback, I added DenseMapInfo specialization for SourceLoc as well. Once we downstream this patch into tensorflow branch, I can change the callsite to use SourceLoc as the key.

Please let me know if there are more comments. Thank you!

**Comment by mhong** [comment]
@swift-ci please test

**Review by jrose-apple — APPROVED** [review]
This seems reasonable to me!

**Comment by mhong** [comment]
Thank you @jrose-apple! I'll attempt a merge.

**Comment by mhong** [comment]
@swift-ci please test and merge

**Comment by mhong** [comment]
@swift-ci please test and merge

**Pull request merged**